File: apps/web/e2e/helpers/gamification.ts
import { Page } from '@playwright/test';

/**
 * Register mock endpoints for gamification scoreboard.
 */
export async function registerGamificationMocks(page: Page) {
  // Stats overview
  await page.route(/\/api\/gamification\/stats/, async (route) => {
    await route.fulfill({
      status: 200,
      contentType: 'application/json',
      body: JSON.stringify({
        success: true,
        data: { totalMembers: 120, totalPointsAwarded: 45000, badgesAwarded: 85, totalEvents: 320 },
      }),
    });
  });

  // Points distribution (pie chart)
  await page.route(/\/api\/gamification\/admin\/points-distribution/, async (route) => {
    await route.fulfill({
      status: 200,
      contentType: 'application/json',
      body: JSON.stringify({
        success: true,
        data: [
          { level: 'Bronze', icon: '🥉', color: '#cd7f32', count: 45 },
          { level: 'Silver', icon: '🥈', color: '#c0c0c0', count: 30 },
          { level: 'Gold', icon: '🥇', color: '#ffd700', count: 18 },
          { level: 'Platinum', icon: '💎', color: '#e5e4e2', count: 8 },
          { level: 'Diamond', icon: '💠', color: '#b9f2ff', count: 4 },
        ],
      }),
    });
  });

  // Points report (top earners table)
  await page.route(/\/api\/gamification\/admin\/points-report/, async (route) => {
    const url = new URL(route.request().url());
    const period = url.searchParams.get('period') || 'weekly';
    const reportData = Array.from({ length: 10 }, (_, i) => ({
      rank: i + 1,
      namaLengkap: `Peserta ${i + 1}`,
      points: Math.round(10000 - i * 800),
      level: [
        'Diamond',
        'Platinum',
        'Gold',
        'Gold',
        'Silver',
        'Silver',
        'Bronze',
        'Bronze',
        'Bronze',
        'Bronze',
      ][i],
      events: Math.round(50 - i * 4),
      lastActive: new Date(2024, 5, 15 - i).toISOString(),
    }));
    await route.fulfill({
      status: 200,
      contentType: 'application/json',
      body: JSON.stringify({
        success: true,
        data:
          period === 'weekly'
            ? reportData
            : reportData.map((r) => ({ ...r, points: r.points * 4 })),
      }),
    });
  });

  // Badges list (gamification main page)
  await page.route(/\/api\/gamification\/badges/, async (route) => {
    await route.fulfill({
      status: 200,
      contentType: 'application/json',
      body: JSON.stringify({
        success: true,
        data: Array.from({ length: 8 }, (_, i) => ({
          id: `badge-${i + 1}`,
          name: [
            'Rajin Latihan',
            'Iuran Tepat Waktu',
            'Prestasi Gemilang',
            'Aktif Organisasi',
            'Pendadaran Sempurna',
            'Mentor Terbaik',
            'Dedikasi Tinggi',
            'Anggota Teladan',
          ][i],
          description: `Badge untuk anggota yang mencapai prestasi ${i + 1}`,
          icon: ['🥋', '💰', '🏆', '👥', '🎓', '🌟', '💪', '⭐'][i],
          threshold: (i + 1) * 10,
          category: [
            'latihan',
            'iuran',
            'prestasi',
            'keaktifan',
            'latihan',
            'keaktifan',
            'prestasi',
            'keaktifan',
          ][i],
        })),
      }),
    });
  });

  // Leaderboard (gamification main page)
  await page.route(/\/api\/gamification\/leaderboard/, async (route) => {
    const url = new URL(route.request().url());
    const skip = parseInt(url.searchParams.get('skip') || '0', 10);
    const limit = parseInt(url.searchParams.get('limit') || '10', 10);
    const data = Array.from({ length: limit }, (_, i) => ({
      rank: skip + i + 1,
      anggotaId: `anggota-${skip + i + 1}`,
      namaLengkap: `Anggota ${skip + i + 1}`,
      points: Math.round(10000 - (skip + i) * 750),
      badges: Math.max(0, 8 - Math.floor((skip + i) / 2)),
      streaks: {
        latihan: Math.max(1, 10 - Math.floor((skip + i) / 3)),
        iuran: Math.max(1, 8 - Math.floor((skip + i) / 4)),
      },
    }));
    await route.fulfill({
      status: 200,
      contentType: 'application/json',
      body: JSON.stringify({
        success: true,
        data,
      }),
    });
  });

  // Events feed (gamification main page)
  await page.route(/\/api\/gamification\/events/, async (route) => {
    await route.fulfill({
      status: 200,
      contentType: 'application/json',
      body: JSON.stringify({
        success: true,
        data: Array.from({ length: 5 }, (_, i) => ({
          id: `event-${i + 1}`,
          anggotaId: `anggota-${i + 1}`,
          namaLengkap: `Anggota ${i + 1}`,
          type: ['training', 'dues', 'badge', 'achievement', 'training'][i],
          points: (i + 1) * 50,
          description: `Mendapatkan ${(i + 1) * 50} poin dari aktivitas ${['latihan', 'iuran', 'badge', 'prestasi', 'latihan'][i]}`,
          timestamp: new Date(2025, 0, 15 - i).toISOString(),
        })),
      }),
    });
  });

  // Org structure (for cascading org filter)
  await page.route(/\/api\/gamification\/org-structure/, async (route) => {
    await route.fulfill({
      status: 200,
      contentType: 'application/json',
      body: JSON.stringify({
        success: true,
        data: [
          {
            id: 'distrik-1',
            nama: 'Distrik Jakarta',
            wilayahs: [
              {
                id: 'wilayah-1',
                nama: 'Jakarta Pusat',
                rantings: [
                  { id: 'ranting-1', nama: 'Menteng' },
                  { id: 'ranting-2', nama: 'Tanah Abang' },
                ],
              },
            ],
          },
          { id: 'distrik-2', nama: 'Distrik Bandung', wilayahs: [] },
        ],
      }),
    });
  });

  // Scoreboard breakdown (module comparison)
  await page.route(/\/api\/gamification\/scoreboard\/breakdown/, async (route) => {
    await route.fulfill({
      status: 200,
      contentType: 'application/json',
      body: JSON.stringify({
        success: true,
        data: [
          { module: 'training', label: 'Latihan', points: 18000, percentage: 40, color: '#3b82f6' },
          { module: 'dues', label: 'Iuran', points: 13500, percentage: 30, color: '#22c55e' },
          { module: 'badge', label: 'Badge', points: 9000, percentage: 20, color: '#eab308' },
          {
            module: 'achievement',
            label: 'Prestasi',
            points: 4500,
            percentage: 10,
            color: '#a855f7',
          },
        ],
      }),
    });
  });
}

File: apps/mobile_flutter/lib/data/models/gamification.dart
/// Model Gamification — memetakan langsung ke JSON dari endpoints
/// `GET /gamification/*` (profile, badges, leaderboard, events, history).
class GamificationLevel {
  final String name;
  final String icon;
  final String color;

  const GamificationLevel({
    required this.name,
    required this.icon,
    required this.color,
  });

  factory GamificationLevel.fromJson(Map<String, dynamic> json) =>
      GamificationLevel(
        name: json['name']?.toString() ?? '',
        icon: json['icon']?.toString() ?? '⭐',
        color: json['color']?.toString() ?? '#16A34A',
      );
}

class GamificationStreaks {
  final int latihan;
  final int iuran;

  const GamificationStreaks({required this.latihan, required this.iuran});

  factory GamificationStreaks.fromJson(dynamic json) {
    if (json is Map<String, dynamic>) {
      return GamificationStreaks(
        latihan: (json['latihan'] as num?)?.toInt() ?? 0,
        iuran: (json['iuran'] as num?)?.toInt() ?? 0,
      );
    }
    return const GamificationStreaks(latihan: 0, iuran: 0);
  }
}

class GamificationBadge {
  final String id;
  final String name;
  final String description;
  final String icon;
  final int threshold;
  final String category;
  final String earnedAt;

  const GamificationBadge({
    required this.id,
    required this.name,
    required this.description,
    required this.icon,
    required this.threshold,
    required this.category,
    this.earnedAt = '',
  });

  factory GamificationBadge.fromJson(Map<String, dynamic> json) =>
      GamificationBadge(
        id: json['id']?.toString() ?? '',
        name: json['name']?.toString() ?? '',
        description: json['description']?.toString() ?? '',
        icon: json['icon']?.toString() ?? '🏅',
        threshold: (json['threshold'] as num?)?.toInt() ?? 0,
        category: json['category']?.toString() ?? '',
        earnedAt: json['earnedAt']?.toString() ?? '',
      );
}

class GamificationProfile {
  final String anggotaId;
  final String namaLengkap;
  final int points;
  final GamificationLevel level;
  final List<GamificationBadge> badges;
  final GamificationStreaks streaks;
  final String lastActivity;

  const GamificationProfile({
    required this.anggotaId,
    required this.namaLengkap,
    required this.points,
    required this.level,
    required this.badges,
    required this.streaks,
    required this.lastActivity,
  });

  factory GamificationProfile.fromJson(Map<String, dynamic> json) {
    final rawLevel = json['level'];
    final rawBadges = json['badges'];
    return GamificationProfile(
      anggotaId: json['anggotaId']?.toString() ?? '',
      namaLengkap: json['namaLengkap']?.toString() ?? '',
      points: (json['points'] as num?)?.toInt() ?? 0,
      level: rawLevel is Map<String, dynamic>
          ? GamificationLevel.fromJson(rawLevel)
          : const GamificationLevel(name: '', icon: '⭐', color: '#16A34A'),
      badges: rawBadges is List
          ? rawBadges
              .whereType<Map<String, dynamic>>()
              .map(GamificationBadge.fromJson)
              .toList()
          : const [],
      streaks: GamificationStreaks.fromJson(json['streaks']),
      lastActivity: json['lastActivity']?.toString() ?? '',
    );
  }
}

class LeaderboardEntry {
  final int rank;
  final String anggotaId;
  final String namaLengkap;
  final int points;
  final int badges;
  final GamificationStreaks streaks;

  const LeaderboardEntry({
    required this.rank,
    required this.anggotaId,
    required this.namaLengkap,
    required this.points,
    required this.badges,
    required this.streaks,
  });

  factory LeaderboardEntry.fromJson(Map<String, dynamic> json) =>
      LeaderboardEntry(
        rank: (json['rank'] as num?)?.toInt() ?? 0,
        anggotaId: json['anggotaId']?.toString() ?? '',
        namaLengkap: json['namaLengkap']?.toString() ?? 'Anggota',
        points: (json['points'] as num?)?.toInt() ?? 0,
        badges: (json['badges'] as num?)?.toInt() ?? 0,
        streaks: GamificationStreaks.fromJson(json['streaks']),
      );
}

class PointEvent {
  final String id;
  final String anggotaId;
  final String namaLengkap;
  final String type;
  final int points;
  final String description;
  final String timestamp;

  const PointEvent({
    required this.id,
    required this.anggotaId,
    required this.namaLengkap,
    required this.type,
    required this.points,
    required this.description,
    required this.timestamp,
  });

  factory PointEvent.fromJson(Map<String, dynamic> json) => PointEvent(
        id: json['id']?.toString() ?? '',
        anggotaId: json['anggotaId']?.toString() ?? '',
        namaLengkap: json['namaLengkap']?.toString() ?? '',
        type: json['type']?.toString() ?? '',
        points: (json['points'] as num?)?.toInt() ?? 0,
        description: json['description']?.toString() ?? '',
        timestamp: json['timestamp']?.toString() ?? '',
      );
}

class PointHistory {
  final String month;
  final int cumulative;

  const PointHistory({required this.month, required this.cumulative});

  factory PointHistory.fromJson(Map<String, dynamic> json) => PointHistory(
        month: json['month']?.toString() ?? '',
        cumulative: (json['cumulative'] as num?)?.toInt() ?? 0,
      );
}

class GamificationReward {
  final String id;
  final String name;
  final String description;
  final String icon;
  final int pointCost;
  final int stock;
  final bool isActive;

  const GamificationReward({
    required this.id,
    required this.name,
    required this.description,
    required this.icon,
    required this.pointCost,
    required this.stock,
    required this.isActive,
  });

  factory GamificationReward.fromJson(Map<String, dynamic> json) =>
      GamificationReward(
        id: json['id']?.toString() ?? '',
        name: json['name']?.toString() ?? '',
        description: json['description']?.toString() ?? '',
        icon: json['icon']?.toString() ?? '🎁',
        pointCost: (json['pointCost'] as num?)?.toInt() ?? 0,
        stock: (json['stock'] as num?)?.toInt() ?? 0,
        isActive: json['isActive'] as bool? ?? true,
      );
}
